fp.go

```diff
@@ -0,0 +1,411 @@
+	"crypto/rand"
+	"fmt"
+	"io"
+	"math/big"
+	"unsafe"
+
+	"github.com/renproject/surge"
+)
+
+// Fp represents an element of the field corresponding to the coordinates of
+// the points that lie on the secp256k1 elliptic curve.
+type Fp struct {
+	inner [5]uint64
+}
+
+// NewFpFromU64 returns a new field element equal to the given unsigned
+// integer.
+func NewFpFromU64(v uint64) Fp {
+	x := Fp{}
+	x.SetU64(v)
+	return x
+}
+
+// SetU64 sets the field element to be equal to the given unsigned integer.
+func (x *Fp) SetU64(v uint64) {
+	// Don't call out to c here as the implementation is simple enough to
+	// warrant avoiding the FFI overhead.
+
+	// Each limb should have no more than 52 nonzero bits when normalized, so
+	// we need to put the lower 52 bits of the argument in the first limb, and
+	// the higher 12 bits in the second limb.
+	x.inner[0] = v & 0xFFFFFFFFFFFFF
+	x.inner[1] = v >> 52
+	x.inner[2] = 0
+	x.inner[3] = 0
+	x.inner[4] = 0
+}
+
+// RandomFp returns a random Fp field element.
+//
+// Panics: This function will panic if there was an error reading bytes from
+// the random source.
+func RandomFp() Fp {
+	x, err := RandomFpNoPanic()
+	if err != nil {
+		panic(fmt.Sprintf("could not generate random bytes: %v", err))
+	}
+	return x
+}
+
+// RandomFpNoPanic returns a random Fp field element or an error if it could
+// not read from the random source.
+func RandomFpNoPanic() (Fp, error) {
+	var bs [32]byte
+	_, err := rand.Read(bs[:])
+	if err != nil {
+		return Fp{}, err
+	}
+	x := Fp{}
+
+	// This will reduce the value modulo P, so it does not matter if the bytes
+	// represent a number greater than P.
+	x.SetB32(bs[:])
+
+	return x, nil
+}
+
+// Clear sets the underlying data of the structure to zero. This will leave it
+// in a state which is a representation of the zero element.
+func (x *Fp) Clear() {
+	// Don't call out to c here as the implementation is simple enough to
+	// warrant avoiding the FFI overhead.
+	x.inner[0] = 0
+	x.inner[1] = 0
+	x.inner[2] = 0
+	x.inner[3] = 0
+	x.inner[4] = 0
+}
+
+// Int returns a big.Int representation of the field element.
+func (x *Fp) Int() *big.Int {
+	ret := new(big.Int)
+	x.PutInt(ret)
+	return ret
+}
+
+// PutInt sets the given big.Int to be equal to the field element.
+func (x *Fp) PutInt(dst *big.Int) {
+	var bs [32]byte
+	x.PutB32(bs[:])
+	dst.SetBytes(bs[:])
+}
+
+// SetB32 sets the field element to be equal to the given byte slice,
+// interepreted as big endian. The field element will be reduced modulo P. This
+// function will return true if the bytes represented a number greater than or
+// equal to P, and false otherwise.
+//
+// Panics: If the byte slice has length less than 32, this function will panic.
+func (x *Fp) SetB32(bs []byte) bool {
+	if len(bs) < 32 {
+		panic(fmt.Sprintf("invalid slice length: length needs to be at least 32, got %v", len(bs)))
+	}
+
+	greater := int(C.secp256k1_fe_set_b32((*C.secp256k1_fe)(unsafe.Pointer(x)), (*C.uchar)(&bs[0]))) == 0
```

Comment:
```suggestion
	greater := int(C.secp256k1_fe_set_b32((*C.secp256k1_fe)(unsafe.Pointer(&x.inner[0])), (*C.uchar)(&bs[0]))) == 0
```

Reply:
There is a stronger guarantee that a pointer to the first element of the contiguous array will be forwards compatible than a pointer to the struct itself (which has no guarantee in Go to not change in future versions, w.r.t. `unsafe.Pointer`).